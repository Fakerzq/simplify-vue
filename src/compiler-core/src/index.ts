/*
 * @Author: Zhouqi
 * @Date: 2022-04-10 14:22:24
 * @LastEditors: Zhouqi
 * @LastEditTime: 2022-04-10 14:22:25
 */
export * from "./compile";
