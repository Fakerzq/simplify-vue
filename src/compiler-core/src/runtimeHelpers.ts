/*
 * @Author: Zhouqi
 * @Date: 2022-04-09 22:40:56
 * @LastEditors: Zhouqi
 * @LastEditTime: 2022-04-24 21:12:47
 */
export const TO_DISPLAY_STRING = Symbol(`toDisplayString`);
export const OPEN_BLOCK = Symbol(`openBlock`);
export const CREATE_ELEMENT_BLOCK = Symbol(`createElementBlock`);
export const CREATE_ELEMENT_VNODE = Symbol(`createElementVNode`);
export const CREATE_TEXT = Symbol(`createTextVNode`);
export const NORMALIZE_CLASS = Symbol(`normalizeClass`);
export const WITH_DIRECTIVES = Symbol(`withDirectives`);

export const helperNameMap: any = {
  [TO_DISPLAY_STRING]: `toDisplayString`,
  [OPEN_BLOCK]: `openBlock`,
  [CREATE_ELEMENT_BLOCK]: `createElementBlock`,
  [CREATE_ELEMENT_VNODE]: `createElementVNode`,
  [CREATE_TEXT]: `createTextVNode`,
  [NORMALIZE_CLASS]: `normalizeClass`,
  [WITH_DIRECTIVES]: `withDirectives`,
};

// 注册运行时辅助函数
export function registerRuntimeHelpers(helpers) {
  Object.getOwnPropertySymbols(helpers).forEach((s) => {
    helperNameMap[s] = helpers[s];
  });
}
