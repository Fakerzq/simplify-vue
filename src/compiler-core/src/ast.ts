/*
 * @Author: Zhouqi
 * @Date: 2022-04-07 22:07:33
 * @LastEditors: Zhouqi
 * @LastEditTime: 2022-04-21 22:05:32
 */
import { CREATE_ELEMENT_BLOCK, CREATE_ELEMENT_VNODE } from "./runtimeHelpers";




export const enum NodeTypes {
  ELEMENT,
  SIMPLE_EXPRESSION,
  INTERPOLATION,
  TEXT,
  ROOT,
  // containers
  COMPOUND_EXPRESSION,
  ATTRIBUTE,
  DIRECTIVE,
  COMMENT,
  VNODE_CALL,
  TEXT_CALL,
  JS_CALL_EXPRESSION,
  JS_PROPERTY,
  JS_OBJECT_EXPRESSION
}

export const enum ElementTypes {
  ELEMENT,
}

// 普通元素的codegenNode
export function createVnodeCall(context, tag, props, children) {
  // context.helper(CREATE_ELEMENT_BLOCK);
  context.helper(CREATE_ELEMENT_VNODE);

  return {
    type: NodeTypes.VNODE_CALL,
    tag,
    props,
    children,
  };
}

// 插值表达式的codegenNode
export function createCallExpression(callee, args) {
  return {
    type: NodeTypes.JS_CALL_EXPRESSION,
    callee,
    arguments: args,
  };
}
