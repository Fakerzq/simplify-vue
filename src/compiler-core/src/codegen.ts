/*
 * @Author: Zhouqi
 * @Date: 2022-04-09 21:13:43
 * @LastEditors: Zhouqi
 * @LastEditTime: 2022-04-10 14:51:48
 */
/**
 * 1. text
 * return function render(_ctx, _cache, $props, $setup, $data, $options) {
 *   return "Hello World";
 * }
 *
 * 2、interplation
 * const { toDisplayString: _toDisplayString } = Vue
 *
 * return function render(_ctx, _cache, $props, $setup, $data, $options) {
 *    return _toDisplayString(_ctx.hello)
 * }
 *
 * 3、element
 * return function render(_ctx, _cache, $props, $setup, $data, $options) {
 *    return (_openBlock(), _createElementBlock("div"))
 * }
 *
 * 4、上面三种结合
 * const { toDisplayString: _toDisplayString, openBlock: _openBlock, createElementBlock: _createElementBlock } = Vue
 *  return function render(_ctx, _cache, $props, $setup, $data, $options) {
 *    return (_openBlock(), _createElementBlock("div", null, "hello, " + _toDisplayString(_ctx.message), 1));
 *  }
 */

import { isFunction, isString } from "../../shared/src";
import { NodeTypes } from "./ast";
import {
  CREATE_ELEMENT_BLOCK,
  helperNameMap,
  TO_DISPLAY_STRING,
} from "./runtimeHelpers";

export function generate(ast, options = {}) {
  const context = createCodegenContext(ast, options);

  genFunctionPreamble(ast, context);

  const { push } = context;
  const functionName = "render";
  const args = ["_ctx", "_cache", "$props", "$setup", "$data", "$options"].join(
    ", "
  );
  push(`function ${functionName}(${args}) { `);
  push(`return `);
  genNode(ast.codegenNode, context);
  push(" }");

  return { code: context.code };
}

/**
 * @author: Zhouqi
 * @description: 创建codegen上下文
 * @param ast
 * @param options
 * @return 上下文对象
 */
function createCodegenContext(ast: any, options: {}) {
  const context = {
    runtimeGlobalName: "Vue",
    // 最终的代码字符串
    code: ``,
    // 字符串拼接操作
    push(text) {
      context.code += text;
    },
    newLine() {
      context.push("\n      ");
    },
    helper(key) {
      return `_${helperNameMap[key]}`;
    },
  };
  return context;
}

/**
 * @author: Zhouqi
 * @description: 根据ast生成节点
 * @param node astNode
 */
function genNode(node, context) {
  const { type } = node;
  const nodeHandlers = {
    [(type === NodeTypes.TEXT) as any]() {
      genText(node, context);
    },
    [(type === NodeTypes.INTERPOLATION) as any]() {
      genInterpolation(node, context);
    },
    [(type === NodeTypes.SIMPLE_EXPRESSION) as any]() {
      genExpression(node, context);
    },
    [(type === NodeTypes.ELEMENT) as any]() {
      genElement(node, context);
    },
    [(type === NodeTypes.COMPOUND_EXPRESSION) as any]() {
      genCompoundExpression(node, context);
    },
  };
  const handler = nodeHandlers[true as any];
  if (isFunction(handler)) {
    handler();
  }
}

/**
 * @author: Zhouqi
 * @description: 生成前导函数代码
 * @param ast
 * @param context
 */
function genFunctionPreamble(ast, context) {
  const { push, runtimeGlobalName, newLine } = context;
  const { helpers } = ast;

  const aliasHelper = (s: symbol) =>
    `${helperNameMap[s]}: _${helperNameMap[s]}`;

  if (helpers.length) {
    push(
      `const { ${helpers.map(aliasHelper).join(", ")} } = ${runtimeGlobalName};`
    );
  }
  newLine();
  push(`return `);
}

/**
 * @author: Zhouqi
 * @description: 生成元素节点的代码字符串
 * @param node
 * @param context
 */
function genElement(node, context) {
  const { push, helper } = context;
  const { tag, children, props } = node;
  push(`${helper(CREATE_ELEMENT_BLOCK)}(`);
  genNodeList(genNullableArgs([tag, props, children]), context);
  push(`)`);
}

function genNullableArgs(args) {
  return args.map((arg) => arg || "null");
}

function genNodeList(nodes, context) {
  const { push } = context;
  for (let i = 0; i < nodes.length; i++) {
    const node = nodes[i];
    if (isString(node)) {
      push(node);
    } else {
      genNode(node, context);
    }
    if (i < nodes.length - 1) {
      push(", ");
    }
  }
}

/**
 * @author: Zhouqi
 * @description: 生成文本节点的代码字符串
 * @param node
 * @param context
 */
function genText(node, context) {
  const { push } = context;
  push(`'${node.content}'`);
}

/**
 * @author: Zhouqi
 * @description: 生成插值节点的代码字符串
 * @param node
 * @param context
 */
function genInterpolation(node, context) {
  const { push, helper } = context;
  push(`${helper(TO_DISPLAY_STRING)}(`);
  genNode(node.content, context);
  push(`)`);
}

/**
 * @author: Zhouqi
 * @description: 生成插值中的表达式的代码字符串
 * @param node
 * @param context
 */
function genExpression(node, context) {
  context.push(node.content);
}

/**
 * @author: Zhouqi
 * @description: 处理复合节点
 * @param node
 * @param context
 */
function genCompoundExpression(node, context) {
  const { children } = node;
  for (let i = 0; i < children.length; i++) {
    const child = children[i];
    if (isString(child)) {
      // +
      context.push(child);
    } else {
      genNode(child, context);
    }
  }
}
