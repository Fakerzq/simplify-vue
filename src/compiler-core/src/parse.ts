/*
 * @Author: Zhouqi
 * @Date: 2022-04-07 21:59:46
 * @LastEditors: Zhouqi
 * @LastEditTime: 2022-04-22 20:25:33
 */
import { extend } from "../../shared/src";
import { ElementTypes, NodeTypes } from "./ast";

// 默认的解析配置
export const defaultParserOptions = {
  delimiters: ["{{", "}}"],
};

/**
 * @author: Zhouqi
 * @description: 模板解析
 * @param template 模板字符串
 * @return ast对象
 */
export function baseParse(template: string) {
  const context = createParserContext(template);
  return createRoot(parseChildren(context, []));
}

/**
 * @author: Zhouqi
 * @description: 创建模板解析上下文对象
 * @param template 模板字符串
 * @return 模板上下文对象
 */
function createParserContext(template: string) {
  const options = extend({}, defaultParserOptions);
  return {
    options,
    source: template,
  };
}

/**
 * @author: Zhouqi
 * @description:
 * @param nodes 模板节点
 * @return ast
 */
function createRoot(nodes) {
  return {
    type: NodeTypes.ROOT,
    children: nodes,
  };
}

/**
 * @author: Zhouqi
 * @description: 解析模板子节点
 * @param context 模板解析上下文对象
 * @param ancestors 存储节点层级关系栈
 * @return 模板子节点
 */
function parseChildren(context, ancestors) {
  const { options } = context;
  const nodes: any = [];
  /**
   * 状态机
   */
  while (!isEnd(context, ancestors)) {
    let node;
    const template = context.source;
    if (template.startsWith(options.delimiters[0])) {
      // 插值节点解析
      node = parseInterpolation(context);
    } else if (template[0] === "<") {
      if (template[1] === "!") {
        if (startsWith(template, "<!--")) {
          // 注释节点
          node = parseComment(context);
        }
      } else if (template[1] === "/") {
        // 结束标签，</xx> 有问题，需要抛出异常
        console.warn("没有开始标签");
        // 截掉这个异常的标签后继续解析
        parseTag(context, TagType.End);
        continue;
      } else if (/[a-z]/i.test(template[1])) {
        // 解析标签
        node = parseElement(context, ancestors);
      }
    }

    // node不存在的话默认就是文本节点
    if (!node) {
      node = parseText(context);
    }
    nodes.push(node);
  }

  return nodes;
}

/**
 * @author: Zhouqi
 * @description: 解析文本节点
 * @param context 模板解析上下文
 * @return 解析后的节点对象
 */
function parseText(context) {
  // 遇到结束截取的标记：
  // 1、结束标签
  // 2、插值模板结束标记
  const endTokens = ["<", context.options.delimiters[0]];
  // 默认截取长度为模板长度
  let endIndex = context.source.length;

  // 遍历模板中的结束标记，找到位置最靠前的结束标记的索引，这个索引就是需要截取的结束位置
  // 例如 <p>div {{ msg }}</p>，此时靠前的结束标记应该是 {{ 而不是 <
  for (let i = 0; i < endTokens.length; i++) {
    const endToken = endTokens[i];
    const index = context.source.indexOf(endToken);
    if (index !== -1 && endIndex > index) {
      endIndex = index;
    }
  }

  const content = parseTextData(context, endIndex);

  return {
    type: NodeTypes.TEXT,
    content,
  };
}

/**
 * @author: Zhouqi
 * @description: 解析注释节点
 * @param context 模板解析上下文
 */
function parseComment(context) {
  const match = /--(\!)?>/.exec(context.source);
  let content;
  if (match) {
    // <!--123--> ===> 123
    content = context.source.slice(4, match.index);
    // TODO 复杂注释
    advanceBy(context, match.index + 3);
  }
  return {
    type: NodeTypes.COMMENT,
    content,
  };
}

/**
 * @author: Zhouqi
 * @description: 获取文本内容和模板推进
 * @param context 模板解析上下文
 * @return 文本内容
 */
function parseTextData(context, length) {
  const content = context.source.slice(0, length);
  advanceBy(context, length);
  return content;
}

/**
 * @author: Zhouqi
 * @description: 解析标签
 * @param context 模板解析上下文
 * @param ancestors 存储节点层级关系栈
 * @return 解析后的节点对象
 */
function parseElement(context, ancestors) {
  const element: any = parseTag(context, TagType.Start);
  // 如果是自闭合标签则直接返回
  if (element.isSelfClosing) return;

  ancestors.push(element);
  // 递归处理子节点
  element.children = parseChildren(context, ancestors);
  ancestors.pop();

  // 判断模板是否合法，比如只有起始标签，没有结束标签，则直接抛出错误
  if (startsWithEndTagOpen(context.source, element.tag)) {
    parseTag(context, TagType.End);
  } else {
    throw "缺少结束标签" + element.tag;
  }
  return element;
}

/**
 * @author: Zhouqi
 * @description: 解析标签
 * @param context 模板解析上下文
 * @param type 标签类型：起始标签/结束标签
 * @return 解析后的标签数据
 */
function parseTag(context, type: TagType) {
  // 正则匹配标签
  const match: any = /^<\/?([a-z][^\t\r\n\f />]*)/i.exec(context.source);
  const tag = match[1];
  // 截掉匹配到的起始内容
  advanceBy(context, match[0].length);
  // 消除无用的空白字符
  advanceSpaces(context);

  const props = parseAttributes(context, type);

  // close tag

  // 是否是自闭合标签 <input />
  let isSelfClosing = false;
  isSelfClosing = startsWith(context.source, "/>");
  // 自闭合标签要截掉2个长度的字符 />，非自闭合截掉一个 >
  advanceBy(context, isSelfClosing ? 2 : 1);

  // 如果是结束标签，直接结束
  if (type === TagType.End) return;

  return {
    // 类型是节点类型
    type: NodeTypes.ELEMENT,
    tag,
    tagType: ElementTypes.ELEMENT,
    isSelfClosing,
    children: [],
    props,
  };
}

/**
 * @author: Zhouqi
 * @description: 解析标签属性
 * @param context 模板解析上下文
 * @param type 标签类型
 * @return 属性
 */
function parseAttributes(context, type) {
  const props: any = [];
  // 存储已经解析过的属性
  const attributeNames = new Set();
  // 当遇到结束标签/自闭合结束标签/模板已经解析完了，则结束解析
  while (
    context.source.length > 0 &&
    !startsWith(context.source, ">") &&
    !startsWith(context.source, "/>")
  ) {
    const attr = parseAttribute(context, attributeNames);
    if (type === TagType.Start) {
      // 起始标签中的属性才需要添加到结果中
      props.push(attr);
    }
    // 截掉空白字符
    advanceSpaces(context);
  }
  return props;
}

/**
 * @author: Zhouqi
 * @description: 解析单个属性
 * @param context 模板解析上下文
 * @param attributeNames 已经解析过的属性集合
 * @return 单个属性解析后的结果
 */
function parseAttribute(context, attributeNames) {
  const match = /^[^\t\r\n\f />][^\t\r\n\f />=]*/.exec(context.source)!;

  // 获取属性名称
  const name = match[0];
  if (attributeNames.has(name)) {
    // 说明出现了重复的属性，需要抛出错误
    console.warn(`the attr ${name} has already defined`);
  }
  attributeNames.add(name);
  // 截掉属性名
  advanceBy(context, name.length);
  // 截掉空白字符
  advanceSpaces(context);
  // 截掉=号
  advanceBy(context, 1);
  // 截掉空白字符
  advanceSpaces(context);

  let value = parseAttributeValue(context);

  // 判断属性是不是v-xxx | @xxx | :xxx | .xxxx | #xxx等vue内部规定的属性定义方式，是的话需要额外进行处理
  if (/^(v-[A-Za-z0-9-]|:|\.|@|#)/.test(name)) {
    const match =
      /(?:^v-([a-z0-9-]+))?(?:(?::|^\.|^@|^#)(\[[^\]]+\]|[^\.]+))?(.+)?$/i.exec(
        name
      )!;

    let dirName;
    if (startsWith(name, ":")) {
      // 属性绑定
      dirName = "bind";
    } else if (startsWith(name, "@")) {
      // 事件绑定
      dirName = "on";
    }

    let arg;
    // match[2]： @click => click :show => show
    if (match[2]) {
      const content = match[2];
      let isStatic = true;

      arg = {
        type: NodeTypes.SIMPLE_EXPRESSION,
        content,
        isStatic,
      };
    }

    return {
      type: NodeTypes.DIRECTIVE,
      name: dirName,
      exp: value && {
        type: NodeTypes.SIMPLE_EXPRESSION,
        content: value.content,
        isStatic: false,
      },
      arg,
    };
  }

  return {
    type: NodeTypes.ATTRIBUTE,
    name,
    value: value && {
      type: NodeTypes.TEXT,
      content: value.content,
    },
  };
}

/**
 * @author: Zhouqi
 * @description: 解析属性值
 * @param context 模板解析上下文
 * @return 属性值
 */
function parseAttributeValue(context) {
  let content;
  const quote = context.source[0];
  const isQuoted = quote === `"` || quote === `'`;
  if (isQuoted) {
    // 截掉起始引号
    advanceBy(context, 1);

    // 如果属性值有引号包裹，则获取引号之间的内容作为属性值
    const endIndex = context.source.indexOf(quote);
    if (endIndex > -1) {
      // 存在结束引号，获取中间内容
      content = parseTextData(context, endIndex);
      // 截掉结束引号
      advanceBy(context, 1);
    } else {
      // 抛出错误
    }
  } else {
    // 处理不是引号包裹的情况
    // 从起始位置到下一个空白字符之前的字符串作为属性值
    const match = /^[^\t\r\n\f >]+/.exec(context.source);
    if (!match) return;
    content = parseTextData(context, match[0].length);
  }
  return {
    content,
    isQuoted,
  };
}

/**
 * @author: Zhouqi
 * @description: 解析插值语法
 * @param context 模板解析上下文
 * @return 解析后的节点对象
 */
function parseInterpolation(context) {
  // template {{ message }}.
  const [openDelimiters, closeDelimiters] = context.options.delimiters;
  const openDelimitersLength = openDelimiters.length;

  const closeDelimitersIndex = context.source.indexOf(
    closeDelimiters,
    openDelimitersLength
  );

  if (closeDelimitersIndex === -1) {
    // 说明没有结束标签
    return;
  }

  // 获取插值内容的长度
  const contentLength = closeDelimitersIndex - openDelimitersLength;
  // 截掉起始插值标签
  advanceBy(context, openDelimitersLength);
  const rawContent = parseTextData(context, contentLength);
  const content = rawContent.trim();
  // 截掉已经解析过的部分
  advanceBy(context, closeDelimiters.length);

  return {
    // 类型是插值节点
    type: NodeTypes.INTERPOLATION,
    content: {
      // 内容是简单表达式
      type: NodeTypes.SIMPLE_EXPRESSION,
      // 非静态节点，也就是内容会变化
      isStatic: false,
      // 插值内容
      content,
    },
  };
}

/**
 * @author: Zhouqi
 * @description: 辅助模板截取（模板推进）
 * @param context 模板解析上下文
 * @param sliceStart 开始截取的位置
 */
function advanceBy(context, sliceStart) {
  context.source = context.source.slice(sliceStart);
}

const enum TagType {
  Start,
  End,
}

/**
 * @author: Zhouqi
 * @description: 模板解析结束的标志
 * @param context 模板解析上下文
 * @param ancestors 根节点标记
 * @param ancestors 存储节点层级关系栈
 * @return 模板是否继续解析的标记
 */
function isEnd(context, ancestors) {
  const template = context.source;
  // 1. 模板字符串开头为结束标签
  if (startsWith(template, "</")) {
    // 当遇到结束标签时去配置之前记录的起始标签，如果匹配到了则结束模板的遍历，避免死循环，例如：<div><span></div>
    for (let i = ancestors.length - 1; i >= 0; i--) {
      const tag = ancestors[i].tag;
      if (startsWithEndTagOpen(template, tag)) {
        return true;
      }
    }
  }

  // 2. 模板字符串为空
  return !template;
}

/**
 * @author: Zhouqi
 * @description: 判断模板是否匹配到某一个结束标签
 * @param source 模板字符串
 * @param tag 标签
 * @return 是否匹配到
 */
function startsWithEndTagOpen(source: string, tag: string): boolean {
  return (
    startsWith(source, "</") &&
    source.slice(2, 2 + tag.length).toLowerCase() === tag.toLowerCase()
  );
}

/**
 * @author: Zhouqi
 * @description: 消费空白字符串
 * @param context 模板解析上下文
 * @return {*}
 */
function advanceSpaces(context): void {
  const match = /^[\t\r\n\f ]+/.exec(context.source);
  match && advanceBy(context, match[0].length);
}

function startsWith(source, target) {
  return source.startsWith(target);
}
