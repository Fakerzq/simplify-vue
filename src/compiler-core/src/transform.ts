/*
 * @Author: Zhouqi
 * @Date: 2022-04-09 20:33:38
 * @LastEditors: Zhouqi
 * @LastEditTime: 2022-04-10 13:53:28
 */
import { isFunction } from "../../shared/src";
import { NodeTypes } from "./ast";
import { TO_DISPLAY_STRING } from "./runtimeHelpers";

/**
 * @author: Zhouqi
 * @description: 转化ast
 * @param root ast
 * @param options 配置
 * @return
 */
export function transform(root, options = {}) {
  const context = createTransformContext(root, options);
  traverseNode(root, context);
  createRootCodegen(root);
  root.helpers = [...context.helpers.keys()];
}

/**
 * @author: Zhouqi
 * @description: 创建codegen所需要的ast
 * @param root ast
 */
function createRootCodegen(root) {
  const child = root.children[0];
  if (child.type === NodeTypes.ELEMENT) {
    root.codegenNode = child.codegenNode;
  } else {
    root.codegenNode = root.children[0];
  }
}

/**
 * @author: Zhouqi
 * @description: 创换上下文对象
 * @param root ast
 * @param options 配置
 * @return 上下文对象
 */
function createTransformContext(root, { nodeTransforms = [] }) {
  const context = {
    root,
    nodeTransforms,
    helpers: new Map(),
    helper(name) {
      context.helpers.set(name, 1);
    },
  };

  return context;
}
/**
 * @author: Zhouqi
 * @description: 转化ast节点
 * @param node ast
 * @param context 上下文对象
 * @return
 */
function traverseNode(node, context) {
  const { type } = node;
  const { nodeTransforms } = context;
  const traverseChildrenType = [NodeTypes.ELEMENT, NodeTypes.ROOT];

  //   if (type === NodeTypes.TEXT) {
  //     node.content = node.content + "123";
  //   }
  const exitFns: any = [];
  for (let i = 0; i < nodeTransforms.length; i++) {
    const onExit = nodeTransforms[i](node, context);
    if (onExit) {
      exitFns.push(onExit);
    }
  }

  const nodeTypeHandlers = {
    [(type === NodeTypes.INTERPOLATION) as any]() {
      context.helper(TO_DISPLAY_STRING);
    },
    [traverseChildrenType.includes(type) as any]() {
      traverseChildren(node.children, context);
    },
  };

  const nodeHandler = nodeTypeHandlers[true as any];
  if (isFunction(nodeHandler)) {
    nodeHandler();
  }
  let i = exitFns.length;
  while (i--) {
    exitFns[i]();
  }
}

/**
 * @author: Zhouqi
 * @description: ast子节点转化
 * @param children ast子节点
 * @param context 上下文对象
 * @return
 */
function traverseChildren(children: [], context) {
  for (let i = 0; i < children.length; i++) {
    traverseNode(children[i], context);
  }
}
