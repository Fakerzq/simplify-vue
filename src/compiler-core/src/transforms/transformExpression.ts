import { NodeTypes } from "../ast";

/*
 * @Author: Zhouqi
 * @Date: 2022-04-09 23:38:10
 * @LastEditors: Zhouqi
 * @LastEditTime: 2022-04-09 23:40:47
 */
export function transformExpression(node, context) {
  if (node.type === NodeTypes.INTERPOLATION) {
    node.content = processExpression(node.content, context);
  }
}

function processExpression(node, context) {
  node.content = `_ctx.${node.content}`;
  return node;
}
