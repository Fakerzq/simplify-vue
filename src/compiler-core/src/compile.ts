/*
 * @Author: Zhouqi
 * @Date: 2022-04-10 14:20:47
 * @LastEditors: Zhouqi
 * @LastEditTime: 2022-04-24 21:33:48
 */
import { extend } from "../../shared/src";
import { generate } from "./codegen";
import { baseParse } from "./parse";
import { transform } from "./transform";
import { transformElement } from "./transforms/transformElement";
import { transformExpression } from "./transforms/transformExpression";
import { transformText } from "./transforms/transformText";
import { transformBind } from "./transforms/vBind";
import { transformOn } from "./transforms/vOn";

export function baseCompile(template, options) {
  const ast = baseParse(template);
  transform(
    ast,
    extend({}, options, {
      nodeTransforms: [transformExpression, transformElement, transformText],
      directiveTransforms: extend({}, options.directiveTransforms, {
        bind: transformBind,
        on: transformOn,
      }),
    })
  );

  return generate(ast);
}
