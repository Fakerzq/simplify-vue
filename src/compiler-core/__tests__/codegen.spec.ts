/*
 * @Author: Zhouqi
 * @Date: 2022-04-09 20:34:26
 * @LastEditors: Zhouqi
 * @LastEditTime: 2022-04-09 23:40:53
 */
import { generate } from "../src/codegen";
import { baseParse } from "../src/parse";
import { transform } from "../src/transform";
import { transformExpression } from "../src/transforms/transformExpression";
describe("Compiler: transform", () => {
  test("codegen string", () => {
    const ast = baseParse(`hello`);

    transform(ast);
    const code = generate(ast);

    expect(code).toMatchSnapshot();
  });

  test("codegen interplation", () => {
    const ast = baseParse(`{{ hello }}`);

    transform(ast, {
      nodeTransforms: [transformExpression],
    });
    const code = generate(ast);

    expect(code).toMatchSnapshot();
  });
});
