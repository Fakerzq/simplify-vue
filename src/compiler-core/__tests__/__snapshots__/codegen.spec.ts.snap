// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compiler: transform context state 1`] = `
Object {
  "code": "const { createElementVNode: _createElementVNode, openBlock: _openBlock, createElementBlock: _createElementBlock } = Vue;
      return function render(_ctx, _cache, $props, $setup, $data, $options) { return (_openBlock(), _createElementBlock(\\"div\\", { class: \\"name\\"} , []))}",
}
`;
